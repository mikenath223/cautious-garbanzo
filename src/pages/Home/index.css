.home-main {
  margin-block: 2.5rem;
}
