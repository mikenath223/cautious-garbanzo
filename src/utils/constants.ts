import { IWeatherInfo } from "./types"

export const initFavCities: string[] = [
  "Tokyo",
  "Delhi",
  "Shanghai",
  "Dhaka",
  "Sao Paulo",
  "Mexico City",
  "Cairo",
  "Beijing",
  "Mumbai",
  "Osaka",
  "Chongqing",
  "Karachi",
  "Kinshasa",
  "Lagos",
  "Istanbul"
]

export const defaultCity: IWeatherInfo = {
  "request": {
    "type": "City",
    "query": "New York, United States of America",
  },
  "location": {
    "name": "New York",
    "country": "United States of America",
    "region": "New York",
    "localtime": "2019-09-07 08:14",
  },
  "current": {
    "observation_time": "12:14 PM",
    "temperature": 13,
    "weather_code": 113,
    "weather_icons": [
      "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
    ],
    "weather_descriptions": [
      "Sunny"
    ],
    "wind_speed": 0,
    "wind_degree": 349,
    "wind_dir": "N",
    "pressure": 1010,
    "precip": 0,
    "humidity": 90,
    "cloudcover": 0,
    "feelslike": 13,
    "uv_index": 4,
    "visibility": 16
  }
}

export const defaultNote = {
  id: "01",
  text: "Welcome"
}
