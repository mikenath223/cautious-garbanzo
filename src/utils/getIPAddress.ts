const fetchIPAddress = async () => {
  try {
    const response = await fetch('https://ipapi.co/json')
    const data = await response.json()
    return data
  } catch (err) {
    console.error(err)
  }
}

export default fetchIPAddress
