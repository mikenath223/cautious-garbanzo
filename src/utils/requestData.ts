import { IWeatherInfo, PromiseFulfilledResult } from "./types"

export const getCityWeather = async (city: string, isCheckCache: boolean): Promise<IWeatherInfo> => {
  try {
    const currentDate = new Date()
    const localStorageKey = `${city.toLowerCase().trim()}-${currentDate.getHours()}-${currentDate.getDate()}-${currentDate.getMonth()}-${currentDate.getFullYear()}`
    const cachedCityData = localStorage.getItem(localStorageKey)
    if (cachedCityData && isCheckCache) {
      const weatherData = JSON.parse(cachedCityData)
      return weatherData
    }
    const url = `http://api.weatherstack.com/current?access_key=04c73e2c81cd37fd215de627927cdd44&query=${city}`;
    const response = await fetch(url)
    const result = await response.json()
    if (result.error) {
      return result?.error?.info
    }
    localStorage.setItem(localStorageKey, JSON.stringify(result))
    return result
  } catch (error) {
    const errMsg = `Failed to retrieve ${city.trim()} data!`
    throw new Error(errMsg)
  }
}

const sortCities = (cities: string[]) => cities.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))

export const getFavsCityWeather = async (cities: string[]) => {
  const fetchRequests = sortCities(cities).map(city => getCityWeather(city, true));

  try {
    const citiesInfo = await Promise.allSettled(fetchRequests)
    const data = citiesInfo.filter(data => data.status === "fulfilled").map(data => (data as PromiseFulfilledResult<IWeatherInfo>).value)
    return data
  } catch (_) {
    return []
  }
}