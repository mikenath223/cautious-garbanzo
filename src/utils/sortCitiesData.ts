import { IWeatherInfo } from "./types";

const sortCitiesData = (cities: IWeatherInfo[] | undefined) => cities?.sort((a, b) => a.location.name.toLowerCase().localeCompare(b.location.name.toLowerCase()))

export default sortCitiesData
