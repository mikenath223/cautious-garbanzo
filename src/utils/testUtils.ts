import { render } from '@testing-library/react';

export const withRender = (component: JSX.Element) =>
  render(component);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => { };
  localStorage.getItem = (key: string) => null;
}