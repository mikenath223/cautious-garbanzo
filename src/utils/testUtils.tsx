import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export const withRender = (component: JSX.Element) =>
  render(<MemoryRouter>{component}</MemoryRouter>);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => {};
  localStorage.getItem = (key: string) => null;
};
