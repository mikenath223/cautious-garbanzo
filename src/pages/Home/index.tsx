import { useNavigate } from "react-router-dom";
import { useCallback, useEffect, useState } from "react";
import FavCities from "../../components/FavCities";
import Searchbar from "../../components/Searchbar";
import useLocationTrack from "../../hooks/useLocationTrack";
import "./index.css";
import { getCityWeather } from "../../utils/requestData";

let isRunGetLocation = false;

export default function Home() {
  const navigate = useNavigate();
  const { latitude, longitude } = useLocationTrack();
  const [error, setError] = useState(false);

  const handleNavigate = useCallback(async () => {
    try {
      const cityWeather = await getCityWeather(
        null,
        false,
        latitude,
        longitude
      );
      if (cityWeather?.name) {
        localStorage.setItem("selected-city", JSON.stringify(cityWeather));
        navigate("/details");
      }
    } catch (error) {
      setError(true);
    }
  }, [latitude, longitude, navigate]);

  const onGetLocation = useCallback(async () => {
    if (latitude && longitude && !isRunGetLocation) {
      isRunGetLocation = true;
      handleNavigate();
    }
  }, [latitude, longitude, handleNavigate]);

  useEffect(() => {
    onGetLocation();
    setError(false);
  }, [onGetLocation]);

  return (
    <main className="max-width-inhibitor home-main">
      <Searchbar />
      {error && (
        <div className="error-search">
          "There was an error retrieving search result! Please try again"
        </div>
      )}
      <FavCities />
    </main>
  );
}
