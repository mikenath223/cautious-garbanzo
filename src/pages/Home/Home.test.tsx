import { runMock, withRender } from "../../utils/testUtils";
import Home from "./index";

runMock();
const mockSuccessResponse = {};
const mockJsonPromise = Promise.resolve(mockSuccessResponse);
const mockFetchPromise = Promise.resolve({
  ok: true,
  json: () => mockJsonPromise,
});

jest.spyOn(localStorage, "setItem");
localStorage.setItem = (key: string, value: string) => {};
localStorage.getItem = (key: string) => null;

describe("Searchbar", function () {
  it("should match snapshot", () => {
    global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
    const { container } = withRender(<Home />);

    expect(container).toMatchSnapshot();
  });
});
