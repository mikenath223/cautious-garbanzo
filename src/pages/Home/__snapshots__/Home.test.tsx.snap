// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar should match snapshot 1`] = `
<div>
  <main
    class="max-width-inhibitor home-main"
  >
    <form>
      <input
        class="search-input"
        id="search"
        name="search"
        placeholder="Search city weather..."
        required=""
        type="text"
        value=""
      />
      <button
        class="submit-btn cursor-pointer"
        type="submit"
      >
        Search
      </button>
    </form>
    <section>
      <ul
        class="favCities"
      />
    </section>
  </main>
</div>
`;
